ShaderGenerators/SingleShaderMaterialGenerator.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System.Text;
+
+namespace SourceGenerators.ShaderGen;
+
+internal class SingleShaderMaterialGenerator
+{
+    private static readonly string[] Namespaces = new string[]
+    {
+        "System",
+        "BaseX",
+        "FrooxEngine",
+        "NEOSPlus.Shaders"
+    };
+
+    private const string ShaderInputPath = @"shaders"; // TODO Use me!
+    private const string ShaderOutputPath = @"parsedShaders";
+
+    public static void Main(string[] args)
+    {
+        Console.WriteLine("Starting...");
+
+        string path = args.Length == 0 ? ShaderOutputPath : args[0];
+        Directory.CreateDirectory(ShaderOutputPath);
+
+        Console.WriteLine($"Got {Path.GetFullPath(path)} as target directory.");
+
+        if (!Directory.Exists(path))
+        {
+            throw new DirectoryNotFoundException($"Directory {path} not found.");
+        }
+
+        int counter = 0;
+        foreach (var file in Directory.EnumerateFiles(path))
+        {
+            if (Path.GetExtension(file) != ".shader")
+            {
+                continue;
+            }
+
+            ParseShader(file);
+            counter++;
+        }
+
+        Console.WriteLine($"Parsed {counter} files");
+    }
+
+    private static void ParseShader(string path)
+    {
+        var parsedShader = UnityShaderParser.Parse(path);
+
+        StringBuilder builder = new StringBuilder();
+
+        foreach (var @namespace in Namespaces)
+        {
+            builder.AppendLine($"using {@namespace};");
+        }
+
+        builder.AppendLine();
+        builder.AppendLine($"[Category(new string[] {{ \"Assets/Materials/NeosPlus/{parsedShader.Name}\"}})]");
+        builder.AppendLine($"public class {parsedShader.Name}Material : SingleShaderMaterialProvider");
```

Comment:
Replace this line with
```cs
builder.AppendLine($"public partial class {parsedShader.Name}Material : SingleShaderMaterialProvider");
```
So that we can override certain methods manually

Reply:
Added in f1eb0286